protocol/README.md

```diff
@@ -4,38 +4,40 @@ This document describes the canonical system architecture used to encrypt and de
 
 ## Architecture
 
-![arch](../diagrams/arch.png)
-
-The canonical architecture contains three major components.
+The canonical architecture contains four major components.
 
 * *TDF Client* - Initiates and drives the TDF encryption and decryption workflows. Only component with access to the content (ciphertext or plaintext).
-* *Entity Attribute Service* (EAS) - Responsible for authenticating TDF Clients and vending them signed [Entity Objects](../schema/EntityObject.md), which act as bearer tokens. The source of truth for attribute provenance and association. A single architecture can have many EAS instance running, for instance, for a TDF Client to join attributes across mulitple orginizations.
-* *Key Access Service* (KAS) - Responsible for authorizing and granting TDF Clients access to rewrapped data key material. If authorized, TDF Clients can use this rewrapped data key to decrypt TDF ciphertext.
+* *OpenID Connect (OIDC) Identity Provider (IdP)* - This system could be any OIDC IdP software.  Virtru has chosen Keycloak as our reference implementation IdP.
+  * From Wikipedia:  "Keycloak is an open source software product to allow single sign-on with Identity and Access Management aimed at modern applications and services. As of March 2018 this JBoss community project is under the stewardship of Red Hat who use it as the upstream project for their RH-SSO product."  Keycloak is open source Apache License 2.0 code.
+  * Any IdP software should work in this scenario provided it can execute some custom code on successful authentication that can:
+    * Read the TDF Client public key from a custom HTTP header sent with the OIDC authentication request.
+    * Construct and send an Attribute Provider web service request, including the public key in the payload.
+    * Return the resulting Claims Object in the signed IdP JWT.
+  * *Virtru Protocol Mapper* is our Keycloak-specific reference implementation of the above functionality.
+* *Attribute Provider* (AP) - A web service that will receive requests from our OIDC IdP custom code (ex: Virtru Protocol Mapper) containing information about authenticated users and return our custom OIDC/OAuth claims in response.  It is the responsibility of Attribute Provider to transform 3rd party IdP claims/metadata to ABAC attributes.  It returns a TDF [Claims Object](../schema/ClaimsObject.md).
+* *Key Access Service* (KAS) - Responsible for authorizing and granting TDF Clients access to rewrapped data key material. If authorized, TDF Clients can use this rewrapped data key to decrypt TDF ciphertext.  A valid OIDC token containing [TDF Claims](../schema/ClaimsObject.md) must be used as a bearer token when communicating with KAS. An otherwise valid and trusted OIDC token without valid TDF Claims will be rejected as a bearer token by KAS.
```

Comment:
Why will a valid OIDC token from a trusted IDP be rejected as a bearer token as opposed to just denying access to a policy due to not having the correct claims?

Reply:
@biscoe916 it will do the latter, which is a more accurate way to describe what happens than `rejected as a bearer token, good point.

Reply:
> Why will a valid OIDC token from a trusted IDP be rejected as a bearer token as opposed to just denying access to a policy due to not having the correct claims?

Good point. I might be wrong but I think there are two scenarios possible. 

- No TDF claims (i.e. no JSON key) in the bearer token KAS gets at all - can happen if you haven't configured Keycloak correctly, so it made sense to call it out somewhere as a misconfiguration.

- TDF claims exist in the bearer token KAS gets (i.e. there is a JSON key in the doc), but they aren't correct according to how policies have been set up for a given org - you don't have a misconfiguration in Keycloak, but your claims are still wrong.

"Rejected as a bearer token" was meant to indicate the first case - the second case would also involve KAS returning an error. Should both cases result in the same error, or should we distinguish?